Issue/3100 drag and drop linked products

AnirudhBhat opened the pull request:
Resolves: #3100

# Description
This PR adds a drag&drop feature for the linked products. Introduces Edit mode in linked product listing screen where drag&drop and deleting a linked product are enabled

# Test cases

#### TC1 Delete option should not be present in linked product listing screen 

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. You shouldn't see the delete icon for the product

#### TC2 Edit overflow menu should be present in linked product listing screen 

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. You should see the `edit` overflow menu

#### TC3 Entering `Edit mode` should hide the `add product` button, show `delete` icon, `drag handler` instead of the product image, and action to finish or cancel the operation

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Once you enter the `Edit mode`, the `Add product` button must disappear, you should now see the `delete` icon and `drag handler` for every product as well as `Checkmark` and `cross` icon in the toolbar


#### TC4 App should maintain `Edit mode` state on the device rotation

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the `edit` option from the overflow menu
3. Rotate the device. The app should maintain the `Edit mode` state

#### TC5 Products in linked listing screen in `Edit mode` can be reordered via drag&drop

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the `edit` option from the overflow menu
3. In `Edit mode` you must be able to reorder products via drag&drop


#### TC6 App should maintain drag&drop state on the device rotation

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the `edit` option from the overflow menu
3. Reorder products and rotate the device. The app should maintain the order on rotation


#### TC7 App should maintain products list state after deleting any product and rotating device

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the `edit` option from the overflow menu
3. Delete any product and rotate the device. The app should maintain the latest product list on rotation

#### TC8 App should maintain reordered products list state when you exit `Edit mode` via `Checkmark` click

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Reorder products click on the `checkmark` icon. The app should maintain the latest product list

#### TC9 App should maintain products list state after deleting any product and when you exit `Edit mode` via `Checkmark` click

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Delete any product and click on the `checkmark` icon. The app should maintain the latest product list

#### TC10 App should undo reorder operation when you exit `Edit mode` via `Cross` icon click

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Reorder products and click the `cross` icon. The app should undo your reorder operation and go back to the original list

#### TC11 App should undo delete operation when you exit `Edit mode` via `Cross` icon click

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Delete any product and click the `cross` icon. The app should undo your delete operation and go back to the original list
 
#### TC12 Reordering the linked products and saving the operation must show the `UPDATE` option in the product detail screen

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Reorder the products and click the `Checkmark` icon.
4. come back to the product detail page and you should see the `UPDATE` option in the toolbar

#### TC13 Deleting any of the linked products and saving the operation must show the `UPDATE` option in the product detail screen

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Delete any product and click the `Checkmark` icon.
4. Come back to the product detail page and you should see the `UPDATE` option in the toolbar
 
#### TC14 Reordering the linked products and canceling the operation must NOT show the `UPDATE` option in the product detail screen

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Reorder the products and click the `cross` icon.
4. come back to the product detail page and you should NOT see the `UPDATE` option in the toolbar

#### TC15 Deleting any of the linked products and canceling the operation must NOT show the `UPDATE` option in the product detail screen

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Delete any product and click the `cross` icon.
4. Come back to the product detail page and you should NOT see the `UPDATE` option in the toolbar

#### TC16 Disable drag and drop for a single product

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Make sure that it has only a single product.
4. App shouldn't allow you to drag and drop

#### TC17 Hide `Edit` option if there are no products

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Delete all of the products and save them.
4. App shouldn't show `Edit` option on the linked products listing screen

#### TC18 Retain order of drag&drop after exiting and coming back to the linked products listing screen

1. Go to the `Linked products listing` screen (Products tab -> Click any product -> Linked Products -> Click on Edit products button either on Upsells or Cross-sells)
2. Enter `Edit mode` by clicking the edit option from the overflow menu
3. Reorder the products and save them
4. Exit from linked product listing screen and come back, you should still see the order being maintrained


# Feature Demo
<table>
  <tr>
    <td>Drag&drop</td>
     <td>Deleting Linked Product</td>
     <td>Night mode</td>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/1331230/116587449-175a7900-a938-11eb-8931-c11675bab73f.gif" width="500" height="600"> </td>
    <td> <img src="https://user-images.githubusercontent.com/1331230/116588708-81275280-a939-11eb-88eb-389fa9871675.gif" width="500" height="600"> </td>
    <td><img src="https://user-images.githubusercontent.com/1331230/116589352-2d693900-a93a-11eb-94ce-e5649487a26a.gif" width=500 height=600></td>
  </tr>
 </table>

#### Note
This PR does not support maintaining the reordering state of linked products after exiting from the app and coming back again, since the backend API does not support sorting of the linked products. However, you can see the sorting order is maintained in the WordPress admin dashboard.


Update release notes:

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AnirudhBhat committed:
Initial commit. Added overflow menu for edit mode

This commit adds overflow menu in GroupedProductListFragment.kt for edit mode

AnirudhBhat committed:
Hides add product button on entering edit mode

This commit Hides add product button in GroupedProductListFragment.kt on entering edit mode

AnirudhBhat committed:
Show delete icon on entering edit mode

This commit hides delete icon by default in GroupedProductListFragment.kt and shows delete icon only on entering edit mode

AnirudhBhat committed:
Add ActionMode while entering edit mode

This commit adds ActionMode while entering edit mode. We can either select back button or done button to exit from ActionMode

AnirudhBhat committed:
Add drag-and-drop functionality to linked product list

This commit adds drag-and-drop functionality to linked product list on entering edit mode. Drag-and-drop functionality is available only in edit mode

AnirudhBhat committed:
Undo user operation on back button click from edit mode

This commit allows user to undo an operation when clicked on back button from ActionMode

AnirudhBhat committed:
maintains state in edit mode on orientation change

This commit maintains state in edit mode on orientation change

AnirudhBhat committed:
Display Edit title in the toolbar in edit mode

AnirudhBhat committed:
Added unit test

This commit adds unit test for restoring previous product list state on back press from edit mode and for linked product swapping logic.

AnirudhBhat committed:
Add drag handler icon on edit mode

This commit adds drag handler icon on edit mode

AnirudhBhat committed:
Consider the product list as changed after drag-and-drop

This commit considers the product list as modified after drag-and-dropping the product list and if it varies from the original linked product list. Doing this enables the update option in product detail screen. Added relevant unit test as well to verify this behaviour

AnirudhBhat committed:
Rename file for better readability

AnirudhBhat committed:
Rename tests for better readability

Renamed few tests for better readability
Changed ActionMode cancelling icon
refactor code around GroupedProductListViewModel.kt

AnirudhBhat committed:
reuse existing method for showing add products button

AnirudhBhat committed:
refactored code in GroupedProductListViewModel.kt

AnirudhBhat committed:
Added highlighter for drag and drop

This commit adds highlighter for drag and drop in GroupedProductListFragment.kt

AnirudhBhat committed:
Used color compatible with night mode for views in edit mode

AnirudhBhat committed:
remove unused code

AnirudhBhat committed:
Fix lint errors

AnirudhBhat committed:
Fix few more lint errors

AnirudhBhat committed:
Refactor code in GroupedProductListViewModel.kt

This commit refactor's code in GroupedProductListViewModel.kt as well as updated and added unit tests

nbradbury commented:
Overall this looks really good, and I like how the "Add product" button is animated when you enter/exit edit mode - that's a very nice touch. I did run into a few issues, though.

I noticed that drag-and-drop is still enabled when there's only a single product, which is ironic because last week we discovered my own code had the same problem (see #3942) :)

![1](https://user-images.githubusercontent.com/3903757/116874889-0c2d7480-abe8-11eb-9986-43e1db293fdd.png)

nbradbury commented:
If you delete the only product and return to the previous screen, the edit menu is still enabled and tapping it takes you to an empty Edit screen.

![2](https://user-images.githubusercontent.com/3903757/116875031-40a13080-abe8-11eb-8e7c-9bb85a8abb76.png)

nbradbury commented:
If you re-order linked products and return to the linked products list, the re-order is retained. However, if you hit the back button then return to the linked products list, the re-order is lost.

![reorder](https://user-images.githubusercontent.com/3903757/116875094-5c0c3b80-abe8-11eb-84b6-7f87df346cb4.gif)

AnirudhBhat committed:
Retain the drag&drop order (Only in-app)

This commit retain's the drag&drop order even after exiting the linked products listing screen. Note that this is only in-app and does not work after force close

AnirudhBhat committed:
Disable drag&drop for single product

AnirudhBhat committed:
Hide edit overflow menu when we have no products

AnirudhBhat committed:
retains the drag&drop order

AnirudhBhat commented:
Thank you for the review @nbradbury. I have addressed all 3 of them and updated the test cases in the PR description as well.

<table>
  <tr>
    <td>Disable drag and drop</td>
     <td>Hide Edit mode</td>
     <td>Retain drag&drop order</td>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/1331230/116900416-56e5d580-ac56-11eb-972b-608eb5cee2ee.gif" width=270 height=480></td>
    <td><img src="https://user-images.githubusercontent.com/1331230/116900532-767cfe00-ac56-11eb-91c4-a94edcd83fb6.gif" width=270 height=480></td>
    <td><img src="https://user-images.githubusercontent.com/1331230/116900616-8dbbeb80-ac56-11eb-852a-b25bb95a7670.gif" width=270 height=480></td>
  </tr>
 </table>

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListFragment.kt:
```diff
@@ -161,6 +161,15 @@ class GroupedProductListFragment : BaseFragment(R.layout.fragment_grouped_produc
+        if (viewModel.getSelectedProductIdsList().isEmpty()) {
+            item.isVisible = false
+        }
+        activity?.invalidateOptionsMenu()
```
Is it necessary to call `invalidateOptionsMenu()` here?

nbradbury replied:
Also, note that `menu.findItem()` can _possibly_ return null. Is there a Kotlin way to handle that?

AnirudhBhat replied:
My bad, makes no sense to call `invalidateOptionsMenu()` here. Also, handled null check for `menu.findItem()`. Thanks for pointing these

nbradbury commented:
@AnirudhBhat Fixes look good! I'll be doing my code review now.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListFragment.kt:
```diff
@@ -70,7 +99,13 @@ class GroupedProductListFragment : BaseFragment(R.layout.fragment_grouped_produc
+                    showAddProductButton(it)
+                    itemTouchHelper.attachToRecyclerView(null)
+                } else {
+                    // Set the UI for edit mode if its edit mode, else disable drag-and-drop
```
Is this comment describing what `setEditModeUI()` does? If so, it seems inaccurate, and perhaps it would be better to comment above the `setEditModeUI()` function instead.

AnirudhBhat replied:
Forgot to edit the comment after refactoring the code :). Fixed this.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListViewModel.kt:
```diff
@@ -161,6 +214,8 @@ class GroupedProductListViewModel @AssistedInject constructor(
     data class GroupedProductListViewState(
         val selectedProductIds: List<Long>,
+        var previouslySelectedProductIds: List<Long>,
+        var isEditMode: Boolean = false, // Indicates whether the screen is in edit mode or not
```
Sorry for the nitpick, but this comment seems unnecessary since the variable name is descriptive.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductItemViewHolder.kt:
```diff
@@ -158,4 +166,8 @@ class ProductItemViewHolder(val viewBinding: ProductListItemBinding) :
             override fun inSelectionHotspot(e: MotionEvent) = true
         }
+
+    fun setEditMode(mode: Boolean) {
```
Using `mode` as a boolean variable name seems a bit out of place. Can something more descriptive be used?

AnirudhBhat replied:
I have used `showEditMode` now. How does this sound?

nbradbury commented on WooCommerce/src/main/res/drawable/ic_close.xml:
```diff
@@ -1,5 +1,5 @@
 <vector android:autoMirrored="true" android:height="18dp"
-    android:tint="@color/color_surface" android:viewportHeight="24.0"
+    android:tint="@color/white" android:viewportHeight="24.0"
```
Is it safe to change this to `@color/white`? It makes sense if `ic_close` is only used for action mode, but I'm wondering if this drawable is used elsewhere and hard-coding the color would affect dark mode? If it is safe to make this change, it might be better to rename the drawable so the color is included (ie: `ic_close_white`).

AnirudhBhat replied:
Thanks for catching this! I made this change initially while working on the task. I was supposed to do it properly before pushing but totally forgot.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListViewModel.kt:
```diff
@@ -88,10 +96,47 @@ class GroupedProductListViewModel @AssistedInject constructor(
-            groupedProductListRepository.getProductList(selectedProductIds)
+            // sort the product list fetched from the database to match the selectedProductIds.
+            // This is done to retain the drag&drop order.
+            groupedProductListRepository.getProductList(selectedProductIds).sortedBy {
```
It looks we have three places where we use `sortedBy` to retain the drag & drop order. Perhaps a separate `fun` could be used for this so the logic is in one place?

nbradbury commented:
@AnirudhBhat I'm done with the first round of code review. Nice work! I didn't find any real problems, I just made a number of nitpicky comments :)

AnirudhBhat committed:
Refactor and address review comments

AnirudhBhat commented:
> @AnirudhBhat I'm done with the first round of code review. Nice work! I didn't find any real problems, I just made a number of nitpicky comments :)

@nbradbury The code is ready for the second round of review. You can see the changes [here](https://github.com/woocommerce/woocommerce-android/pull/3940/commits/794735ae6e80e2da1691953b80a33271dc94dc59) :)

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/GroupedProductListFragment.kt:
```diff
@@ -237,6 +237,7 @@ class GroupedProductListFragment : BaseFragment(R.layout.fragment_grouped_produc
         }
         viewModel.setEditMode(false)
         viewModel.isActionModeClicked = false
+        requireActivity().invalidateOptionsMenu()
```
👍

nbradbury commented:
@AnirudhBhat Your changes look good! I'll give my feedback to the technical recruiter and get back to you. Again, nice work on this!

nbradbury commented:
Closing this without merging since the backend doesn't support retaining the order of linked products :(
